effectful/handlers/llm/providers.py

```diff
@@ -422,13 +455,18 @@ def decode_response[**P, T](template: Callable[P, T], response: ModelResponse) -
     assert result_str
 
     ret_type = template.__signature__.return_annotation
-    if ret_type == str:
-        return result_str  # type: ignore[return-value]
+    encodable_ty = type_to_encodable_type(ret_type)
 
-    Result = _pydantic_model_from_type(ret_type)
-    result = Result.model_validate_json(result_str)
-    assert isinstance(result, Result)
-    return result.value
+    if encodable_ty.t == str:
```

Comment:
As mentioned above, it would be nice to encapsulate any special-casing for `str` within `type_to_encodable_type`, rather than having it appear in downstream code like this.